### DISCUSSION TITLE
How to add type safety to a store in implementation

### DISCUSSION THREAD

**Discussion by lkdm** [pmndrs/zustand#1022] [Q&A]
We are using Zustand as a global store for our application state. Part of this is that I want to have a slice of store called *UIState* which holds state for the current page (or state of UI). Ideally, I would like this slice to be type-safe.

The slice is created as part of our global store, but I'd ideally like the interface for it to come from the PageComponent I'm using it with.

## uislice.tsx
```
export interface UIState {
    // Parts of UI State that are common across all pages.
    isLoading: boolean,
    alert?: Alert
    pagination?: PaginationSettings,
}

interface UISlice {
    ui: UIState,

    // Some actions
    setUI: (ui: UIState) => void,
    setIsLoading: (value: boolean) => void,
    setAlert: (alert: Alert) => void,
    clearAlert: () => void,
    setPagination: (pagination: PaginationSettings) => void
}

const createUISlice: StoreSlice<UISlice> = (set: SetState<UISlice>) => ({
    ui: {
        isLoading: true,
        alert: undefined,
        pagination: undefined
    },
    ...uiStateActions
```

I want to extend `UIState` 

```
interface MyUiState extends UIState {
    employeesById: Array<UUID>,
    summary: WorkloadSummary,
}
```

And use it to instantiate that part of store.

```
// In the component
const store = useStore<MyUiState>(state => state.ui);
```

But I'm getting a Typescript error

```
Property 'employeesById' does not exist on type 'UIState'
```

I know why it's doing this. I haven't actually instantiated the state with an initial object in the shape of `MyUiState`. I'm wondering if there's a better way of doing this, than manually setting it in the component.

I and the devs I work with love Zustand btw 🙏

**Comment by lkdm** [comment]
I think I may have been misguided with trying to use `state.ui` as a store for each PageComponent. Maybe I'm better off just having a separate store, created in the component for that.

**Comment by dai-shi** [comment]
> I know why it's doing this. I haven't actually instantiated the state with an initial object in the shape of `MyUiState`.

If you are doing something that only you can know, you would end up using type assertion, I guess.

**Reply by lkdm** [reply]
Thanks @dai-shi, I think I'm better of just using a function to create a new local store for each component.
